contracts/mainnet/connectors/uniswapStaker/main.sol

```diff
@@ -0,0 +1,214 @@
+        address _refundee,
+        uint256 _tokenId
+    )
+        external
+        payable
+        returns (string memory _eventName, bytes memory _eventParam)
+    {
+        address poolAddr = getPoolAddress(_tokenId);
+
+        IUniswapV3Pool pool = IUniswapV3Pool(poolAddr);
+        IUniswapV3Staker.IncentiveKey memory _key = IUniswapV3Staker
+            .IncentiveKey(
+                IERC20Minimal(_rewardToken),
+                pool,
+                _startTime,
+                _endTime,
+                _refundee
+            );
+        _stake(_tokenId, _key);
+
+        _eventName = "LogStake(uint256, address)";
+        _eventParam = abi.encode(_tokenId, _refundee);
+    }
+
+    /**
+     * @dev Unstake NFT LP token
+     * @notice Unstake NFT LP Position
+     * @param _rewardToken _rewardToken address
+     * @param _startTime stake start time
+     * @param _endTime stake end time
+     * @param _refundee refundee address
+     * @param _tokenId NFT LP token id
+     */
+    function unstake(
+        address _rewardToken,
+        uint256 _startTime,
+        uint256 _endTime,
+        address _refundee,
+        uint256 _tokenId
+    )
+        external
+        payable
+        returns (string memory _eventName, bytes memory _eventParam)
+    {
+        address poolAddr = getPoolAddress(_tokenId);
+
+        IUniswapV3Pool pool = IUniswapV3Pool(poolAddr);
+        IUniswapV3Staker.IncentiveKey memory _key = IUniswapV3Staker
+            .IncentiveKey(
+                IERC20Minimal(_rewardToken),
+                pool,
+                _startTime,
+                _endTime,
+                _refundee
+            );
+        _unstake(_key, _tokenId);
+        _eventName = "LogUnstake(uint256,bytes32)";
+        _eventParam = abi.encode(_tokenId, _key);
+    }
+
+    /**
+     * @dev Claim rewards
+     * @notice Claim rewards
+     * @param _rewardToken _rewardToken address
+     * @param _to address to receive
+     * @param _amountRequested requested amount
+     */
+    function claimRewards(
+        address _rewardToken,
+        address _to,
+        uint256 _amountRequested
+    )
+        external
+        payable
+        returns (string memory _eventName, bytes memory _eventParam)
+    {
+        uint256 rewards = _claimRewards(
+            IERC20Minimal(_rewardToken),
+            _to,
+            _amountRequested
+        );
+
+        _eventName = "LogRewardClaimed(address,address,uint256)";
+        _eventParam = abi.encode(_rewardToken, _to, rewards);
+    }
+
+    /**
+     * @dev Create incentive
+     * @notice Create incentive
+     * @param _rewardToken _rewardToken address
+     * @param _length incentive length
+     * @param _refundee refundee address
+     * @param _tokenId NFT LP token id
+     * @param _reward reward amount
+     */
+    function createIncentive(
+        address _rewardToken,
+        uint256 _length,
+        address _refundee,
+        uint256 _tokenId,
```

Comment:
I used _tokenId to get uniswap pool.

Reply:
@cryptoDev222 Let's pass poolAddress directly?

Reply:
@cryptoDev222 After creating incentive, when should we deposit reward tokens or how should we?

Reply:
Because I am not able to see any approve function for this: https://github.com/Uniswap/uniswap-v3-staker/blob/eff32b5509f87b3e66734c42ba5fc7c607dc583c/contracts/UniswapV3Staker.sol#L117

Reply:
The rewards tokens are deposited when we create incentive, so we should approve first, before we create incentive.
Okay, I'll pass poolAddress directly.

Reply:
Sure, we could implement approve method.